# Assignment #2: Data Exploration

In this assignment, you will demonstrate your knowledge of the Python and Pandas skills we've learned so far. These include:

- Getting an overview of your data
- Extracting columns
- Removing duplicates from your data
- Creating a subset of your data by matching a string
- Getting a random sample of your data
- Sorting your data 
- Getting fundamental statistical information of a series, such as the mean and median

You will not need to formulate or answer a research question for this assignment. You simply need to demonstrate your ability to perform specific operations on the data set you have selected. In the next assignment, your midterm, you will create a narrative exploration of the data using many of the same methods, but in such a way as to use narrative along with data analysis to answer a research question.

You may use the data set you imported in assignment #1, or you may choose another data set. That means that you can reuse code you wrote in Assignment #1 to import the data.

In the below notebook, I will ask for you to perform a task. Each task will be followed by one or more blank code cells. Use as many cells as you need to complete the task. If you know some other way to complete the task with Python and/or Pandas that is not the way we learned in class, that is also acceptable. Feel free to use Google searches to refresh your memory as to how to perform each task. You will also likely wish to consult the class notebooks, which can be found in [our class repository on GitHub](https://github.com/sha256rma/foundations-of-data-science).

If your data set does not have a needed form of data for a task, write code that would perform the task if a column with that data existed, and provide an explanation as a comment or markdown cell.

You may receive partial credit for some incomplete or incorrect answers. Please feel free to add comments about your thought process, which will make it more likely that you will receive partial credit.

The last cell is a bonus. You will not be penalized for not attempting or completing it.

-----

Import Pandas and read your data in as a data frame. Assign the data frame to a variable, such as `df`.

In [762]:
import pandas as pd

In [764]:
data = pd.read_csv('ProductPriceIndex.csv')

Make the data frame visible by placing the data frame variable in a cell by itself.

In [767]:
data

productname        date  ... newyorkretail averagespread
0          Strawberries  2019-05-19  ...         $2.54        82.33%
1       Romaine Lettuce  2019-05-19  ...         $1.99       428.57%
2      Red Leaf Lettuce  2019-05-19  ...         $1.89       467.19%
3              Potatoes  2019-05-19  ...         $6.22       244.50%
4               Oranges  2019-05-19  ...         $2.05       281.71%
...                 ...         ...  ...           ...           ...
15761   Broccoli Crowns  1999-10-24  ...         $1.15       208.62%
15762  Broccoli Bunches  1999-10-24  ...         $1.10       242.39%
15763       Cauliflower  1999-10-24  ...         $0.92       252.63%
15764           Carrots  1999-10-24  ...         $0.59       331.25%
15765        Cantaloupe  1999-10-24  ...         $0.79       125.00%

[15766 rows x 8 columns]

What columns are in your data frame? Use a function or method that shows all the columns.

In [770]:
data.columns

Index(['productname', 'date', 'farmprice', 'atlantaretail', 'chicagoretail',
       'losangelesretail', 'newyorkretail', 'averagespread'],
      dtype='object')

Write Python code to output the number of rows in your data frame.

In [773]:
len(data)

15766

Write code to output a column from your data frame as a series. (Extract a column.)

In [776]:
series_one = pd.Series(data.productname)


In [778]:
series_one

0            Strawberries
1         Romaine Lettuce
2        Red Leaf Lettuce
3                Potatoes
4                 Oranges
               ...       
15761     Broccoli Crowns
15762    Broccoli Bunches
15763         Cauliflower
15764             Carrots
15765          Cantaloupe
Name: productname, Length: 15766, dtype: object

### Read the example below to **count** duplicates based on specific columns (e.g., 'Name' and 'Age')

Check duplicates based on 'Name' and 'Age' columns (modify the subset for your own columns):

`duplicates_subset = df.duplicated(subset=['Name', 'Age'])`

Count the number of duplicates in the subset:

`num_duplicates_subset = duplicates_subset.sum()`

Display no. of duplicate rows based on your column(s):

`print(f"\nNumber of duplicate rows based on 'Name' and 'Age': {num_duplicates_subset}")`


After reading the above, modify the code snippets to check for duplicate columns in your own data frame

In [780]:
data

productname        date  ... newyorkretail averagespread
0          Strawberries  2019-05-19  ...         $2.54        82.33%
1       Romaine Lettuce  2019-05-19  ...         $1.99       428.57%
2      Red Leaf Lettuce  2019-05-19  ...         $1.89       467.19%
3              Potatoes  2019-05-19  ...         $6.22       244.50%
4               Oranges  2019-05-19  ...         $2.05       281.71%
...                 ...         ...  ...           ...           ...
15761   Broccoli Crowns  1999-10-24  ...         $1.15       208.62%
15762  Broccoli Bunches  1999-10-24  ...         $1.10       242.39%
15763       Cauliflower  1999-10-24  ...         $0.92       252.63%
15764           Carrots  1999-10-24  ...         $0.59       331.25%
15765        Cantaloupe  1999-10-24  ...         $0.79       125.00%

[15766 rows x 8 columns]

In [802]:
"""
(0, productname         Strawberries
date                  2019-05-19
farmprice                  $1.16
atlantaretail              $2.23
chicagoretail              $1.70
losangelesretail           $1.99
newyorkretail              $2.54
averagespread             82.33%
Name: 0, dtype: object)
"""

for row in data.iterrows():
    farm_price = row[1].farmprice
    new_york_retail =  row[1].newyorkretail
    chicago_retail = row[1].chicagoretail
    if farm_price == new_york_retail == chicago_retail:
        data.drop(row[0])

If any duplicates exist within a column, remove them with the .drop_duplicates() function

In [806]:
data_dropped_productname = data.drop_duplicates(subset='productname')

Write Python code to output the number of rows in your data frame.

In [835]:
data_dropped_productname

productname        date  ... newyorkretail averagespread
0          Strawberries  2019-05-19  ...         $2.54        82.33%
1       Romaine Lettuce  2019-05-19  ...         $1.99       428.57%
2      Red Leaf Lettuce  2019-05-19  ...         $1.89       467.19%
3              Potatoes  2019-05-19  ...         $6.22       244.50%
4               Oranges  2019-05-19  ...         $2.05       281.71%
5       Iceberg Lettuce  2019-05-19  ...         $1.56       335.71%
6    Green Leaf Lettuce  2019-05-19  ...         $1.71       450.00%
7                Celery  2019-05-19  ...         $2.49       -18.20%
8           Cauliflower  2019-05-19  ...         $2.97       198.98%
9               Carrots  2019-05-19  ...         $1.22       328.13%
10           Cantaloupe  2019-05-19  ...         $2.94       120.40%
11      Broccoli Crowns  2019-05-19  ...         $1.99       148.97%
12             Avocados  2019-05-19  ...         $1.69        30.49%
25     Broccoli Bunches  2019-05-12  ...         $1.48        42.89%
78            Asparagus  2019-04-14  ...         $1.61       -12.96%
264        Flame Grapes  2018-12-30  ...         $2.71       167.50%
270     Thompson Grapes  2018-12-23  ...         $2.59       126.33%
348           Honeydews  2018-11-18  ...         $4.61       141.27%
398            Tomatoes  2018-10-21  ...         $1.74       129.69%
503               Plums  2018-09-09  ...         $2.49       188.03%
504             Peaches  2018-09-09  ...         $2.49       204.69%
522          Nectarines  2018-09-02  ...         $2.49       212.50%

[22 rows x 8 columns]

Compare the new value with the first time you outputted the number of rows above. Is it more, less, or the same?

First_value = 15766 > Second_value = 22 rows

In [812]:
filter = data['productname'] == 'Strawberries'
subset = data.loc[filter]

In [814]:
subset

productname        date  ... newyorkretail averagespread
0      Strawberries  2019-05-19  ...         $2.54        82.33%
13     Strawberries  2019-05-12  ...         $2.66       166.21%
27     Strawberries  2019-05-05  ...         $2.91       302.59%
39     Strawberries  2019-04-28  ...         $2.96       334.48%
52     Strawberries  2019-04-21  ...         $3.16       280.43%
...             ...         ...  ...           ...           ...
15657  Strawberries  1999-12-26  ...         $3.62       118.20%
15709  Strawberries  1999-11-21  ...         $3.34       113.52%
15732  Strawberries  1999-11-14  ...         $3.34       113.52%
15741  Strawberries  1999-11-07  ...         $3.16       103.07%
15752  Strawberries  1999-10-31  ...         $2.66       140.77%

[973 rows x 8 columns]

Output the first ten rows of your data set, the last ten rows of your data set, and a random ten rows of your data set (a sample of your data).

In [840]:
print(data.head(10))

          productname        date  ... newyorkretail averagespread
0        Strawberries  2019-05-19  ...         $2.54        82.33%
1     Romaine Lettuce  2019-05-19  ...         $1.99       428.57%
2    Red Leaf Lettuce  2019-05-19  ...         $1.89       467.19%
3            Potatoes  2019-05-19  ...         $6.22       244.50%
4             Oranges  2019-05-19  ...         $2.05       281.71%
5     Iceberg Lettuce  2019-05-19  ...         $1.56       335.71%
6  Green Leaf Lettuce  2019-05-19  ...         $1.71       450.00%
7              Celery  2019-05-19  ...         $2.49       -18.20%
8         Cauliflower  2019-05-19  ...         $2.97       198.98%
9             Carrots  2019-05-19  ...         $1.22       328.13%

[10 rows x 8 columns]


In [842]:
print(data.tail(10))

              productname        date  ... newyorkretail averagespread
15756    Red Leaf Lettuce  1999-10-31  ...         $1.29       439.71%
15757            Potatoes  1999-10-31  ...         $2.72       161.86%
15758     Iceberg Lettuce  1999-10-31  ...         $1.19       327.50%
15759           Honeydews  1999-10-31  ...         $0.82       212.50%
15760  Green Leaf Lettuce  1999-10-31  ...         $1.29       439.71%
15761     Broccoli Crowns  1999-10-24  ...         $1.15       208.62%
15762    Broccoli Bunches  1999-10-24  ...         $1.10       242.39%
15763         Cauliflower  1999-10-24  ...         $0.92       252.63%
15764             Carrots  1999-10-24  ...         $0.59       331.25%
15765          Cantaloupe  1999-10-24  ...         $0.79       125.00%

[10 rows x 8 columns]


In [844]:
print(data.sample(10))

              productname        date  ... newyorkretail averagespread
4282     Red Leaf Lettuce  2013-10-20  ...         $1.86       377.86%
11808          Nectarines  2004-09-05  ...         $1.67       417.14%
6316   Green Leaf Lettuce  2011-04-10  ...         $0.90       305.00%
14449             Carrots  2001-07-01  ...         $0.82       319.12%
9727         Strawberries  2007-03-11  ...         $2.99        90.63%
12395     Broccoli Crowns  2003-12-28  ...         $1.67       101.23%
3763              Peaches  2014-06-29  ...         $0.00        20.21%
7388      Romaine Lettuce  2009-11-22  ...         $2.24        70.67%
7506      Romaine Lettuce  2009-10-04  ...         $1.07       439.00%
12830    Red Leaf Lettuce  2003-07-06  ...         $1.79       655.56%

[10 rows x 8 columns]


## Bonus

You may want to refer to the [Pandas Cheat Sheet](https://pandas.pydata.org/Pandas_Cheat_Sheet.pdf).

Sort your data frame by a particular column and output the result.

Pick a column in your data set with numeric data (integers or floats). Output the mean (average) and median of that column.